x-pack/plugins/security_solution/server/lib/entity_analytics/asset_criticality/asset_criticality_migration_client.ts

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import type { Logger, ElasticsearchClient } from '@kbn/core/server';
+import type { AuditLogger } from '@kbn/security-plugin-types-server';
+import { AssetCriticalityDataClient } from './asset_criticality_data_client';
+
+interface AssetCriticalityEcsMigrationClientOpts {
+  logger: Logger;
+  auditLogger: AuditLogger | undefined;
+  esClient: ElasticsearchClient;
+}
+
+const ECS_MAPPINGS_MIGRATION_QUERY = {
+  bool: {
+    must_not: {
+      exists: {
+        field: 'asset.criticality',
+      },
+    },
+  },
+};
+
+export class AssetCriticalityEcsMigrationClient {
+  private readonly assetCriticalityDataClient: AssetCriticalityDataClient;
+  constructor(private readonly options: AssetCriticalityEcsMigrationClientOpts) {
+    this.assetCriticalityDataClient = new AssetCriticalityDataClient({
+      ...options,
+      namespace: '*', // The migration is applied to all spaces
+    });
+  }
+
+  public isEcsMappingsMigrationRequired = async () => {
+    const indicesMappings = await this.assetCriticalityDataClient.getIndexMappings();
+
+    return Object.values(indicesMappings).some(
+      ({ mappings }) => mappings?.properties?.asset === undefined
+    );
+  };
+
+  public isEcsDataMigrationRequired = async () => {
+    const resp = await this.assetCriticalityDataClient.search({
+      query: ECS_MAPPINGS_MIGRATION_QUERY,
+      size: 1,
+    });
+
+    return resp.hits.hits.length > 0;
+  };
+
+  public migrateEcsMappings = () => {
+    return this.assetCriticalityDataClient.createOrUpdateIndex();
+  };
+
+  public migrateEcsData = () => {
+    return this.options.esClient.updateByQuery(
+      {
+        index: this.assetCriticalityDataClient.getIndex(),
+        conflicts: 'proceed',
+        ignore_unavailable: true,
+        allow_no_indices: true,
+        scroll_size: 1000,
+        body: {
+          query: ECS_MAPPINGS_MIGRATION_QUERY,
+          script: {
+            source: `
+          Map asset = new HashMap();
```

Comment:
maybe extract this into a constant as well? I recall there seemed to be a way to import .painless files too?

Reply:
The script that reads `.painless` is tailor-made for the risk engine job.
I will move it to a constant. Hopefully, it will have low maintenance because migration scripts can't change.